repos/base-hw/src/bootstrap/spec/arm_64/crt0.s

```diff
@@ -19,11 +19,33 @@
 	and x0, x0, #0b11111111
 .endm
 
+/**
+ * Disable MMU for EL1 and EL2 if applicable.
+ */
+.macro _mmu_disable
+	mrs  x8, CurrentEL
+	lsr  x8, x8, #2
+	cmp  x8, #0x2
+	b.lo 1f
+
+	mrs x8, sctlr_el1
+	bic x8, x8, #(1 << 0)
+	msr sctlr_el1, x8
+	isb
+1:
+	mrs x8, sctlr_el2
+	bic x8, x8, #(1 << 0)
+	msr sctlr_el2, x8
+	isb
+.endm
+
```

Comment:
@tworaz I highly appreciate your initiative to be more robust in different boot environments. However, I wonder whether we then would have to check for EL3 too. Moreover, isn't it possible to move this code sequence into high-level C++, e.g., at the beginning of the processor-specific enable_mmu functionality? I would expect all code in between to not interfer with being executed with enabled MMU. For instance on x86, we also typically get booted with 1:1 mappings and enabled MMU. But maybe I oversee some problems here?!
It is just that less people speak ARMv8 assembler, and such branching is better readable in C++ than in assembler. So if it is possible to disable the MMU later, I would favor it.

Reply:
I'll have to try if it works. I did originally try to disable MMU at the beginning of enable_mmu, but it did not work. The primary core booted fine, but secondary ones still got stuck somewhere around ttbr write calls. At least it seemed like it since I don't have any good debugger that would allow me to pinpoint the exact location where things break.

The thing is this happened before I discovered the problem addressed by the second patch. It might be that disabling MMU at the beginning of enable_mmu call will work fine now. Without the MPIDR patch all the cores where considered as core 0 and used the same stack from crt0.s.

I should have some time in the evening to run some tests.

Reply:
@skalk So I run some tests yesterday and things are not so simple. Moving this code into C++ enable_mmu function did not work. At least not for SMP use case. With mmu disabled at the beginning of enable_mmu function things boot fine with NR_OF_CPUS=1. With NR_OF_CPUS>1 all the secondary cores get stuck somewhere before enable_mmu. They do seem to get into bootstrap's init function, but not further than that.

So next I've tried experimenting with the _mmu_disable call this patch adds in crt0.s. What I've managed to establish is the mmu needs to be disabled for primary core before the _crt0_fill_bss_zero section in this file. If the primary core disables MMU after clearing BSS, secondary cores will not start as they should. Secondary cores themselves can disable MMU at the beginning of enable_mmu C++ function and things will continue to work.

Right now I'm not really sure what to make of it. This smells like some sort of problem with dcache. Also it might be that u-boot for quartz64 does not actually setup identity mappings for the MMU. If it did that I'd expect for primary core to also fail to boot.

Reply:
@tworaz thank you for investigating the problem again. I see that it is problematic if the boot environment did not consistently setup the different cores. So I agree with disabling the MMU within the assembler path. However, after looking again at the commit, I've seen that you unconditionally disable the MMU of EL2, also if you detected EL1, am I right? I would think this will cause a fault, doesn't it? I'll rework your commit slightly to address EL1-EL3 and only disable the MMU of the actual exception level. Everything else can be addressed at the higher levels, when preparing less privileged levels.

Reply:
@skalk  You're right the patch I've pushed has EL2/EL1 branches swapped. I made this mistake originally when cleaning up the code. Then I found the bug, made a fixup, but apparently this fixup targeted the wrong patch. The patch I cheery picked to this review did not have that fixup since it was squashed into another patch. I've just pushed the correct the patch.

The goal is to disable MMU for both EL1 and EL2 when booting in EL2 and only in EL1 when starting in this exception level. I can do something similar for EL3, but I have no HW to actually test it on. If you want I can only disable MMU for the exception level in which the code boots and handle higher levels in C++ code.

If you have the time feel free to rework this patch to something you feel is better suited for integrating into main Genode branch. I probably won't have the time to work on Genode related stuff in the next 2 days.